gadget/filesystemimage_test.go

```diff
@@ -0,0 +1,375 @@
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, "post stage callback failed: post stage exploded")
+}
+
+func (s *filesystemImageTestSuite) TestMkfsError(c *C) {
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("will not mkfs")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, s.psTrivial, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, s.psTrivial)
+
+	err = fiw.Write(img, nil)
+	c.Assert(err, ErrorMatches, `cannot create "happyfs" filesystem: will not mkfs`)
+}
+
+func (s *filesystemImageTestSuite) TestFilesystemExtraCheckError(c *C) {
+	ps := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Filesystem: "happyfs",
+			Size:       2 * gadget.SizeMiB,
+			Content:    []gadget.VolumeContent{},
+		},
+	}
+
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("will not mkfs")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, ps, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, ps)
+
+	// modify filesystem
+	ps.Filesystem = "foofs"
+
+	err = fiw.Write(img, nil)
+	c.Assert(err, ErrorMatches, `internal error: filesystem "foofs" has no handler`)
+}
+
+func (s *filesystemImageTestSuite) TestMountedWriterError(c *C) {
+	ps := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Filesystem: "happyfs",
+			Size:       2 * gadget.SizeMiB,
+			Content: []gadget.VolumeContent{
+				{Source: "/foo", Target: "/"},
+			},
+		},
+	}
+
+	cb := func(rootDir string, cbPs *gadget.PositionedStructure) error {
+		return errors.New("unexpected call")
+	}
+
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("unexpected call")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
+		"happyfs": mkfsHappyFs,
+	})
+	defer restore()
+
+	fiw, err := gadget.NewFilesystemImageWriter(s.content, ps, s.work)
+	c.Assert(err, IsNil)
+
+	img := s.imgForPs(c, ps)
+
+	// declared content does not exist in the content directory
+	err = fiw.Write(img, cb)
+	c.Assert(err, ErrorMatches, `cannot prepare filesystem content: cannot write filesystem content of source:/foo: .* no such file or directory`)
+}
+
+func (s *filesystemImageTestSuite) TestBadWorkDirError(c *C) {
+	cb := func(rootDir string, cbPs *gadget.PositionedStructure) error {
+		return errors.New("unexpected call")
+	}
+	mkfsHappyFs := func(imgFile, label, contentsRootDir string) error {
+		return errors.New("unexpected call")
+	}
+	restore := gadget.MockMkfsHandlers(map[string]gadget.MkfsFunc{
```

Comment:
(nitpick^2) Given how often this is called I wonder if it would make sense to have it in SetUpTest (or maybe have a sub-suite that does it?).

Reply:
Went further. Added a sub suite with mocks for `happyfs`, as well as another sub-suite that verifies the default handlers for `ext4` and `vfat`.